### ISSUE TITLE
Protocol 19 - message 0x45 - minimap

### ISSUE THREAD

**Issue by Barbosik** [issy123/agario-protocol#22]
Since 2018/20/12, a new protocol 19 is rolled out. The difference is just in cipher key. 
Also added a new message 0x45 - Minimap

sv_protocol 19
sv_secret 0x0000765D
sv_versionClient "3.3.1"
sv_versionProto "12.0.1"

**Comment by Barbosik** [comment]
45 = MINIMAP UPDATE

| Offset | Size | Description |
| --- | --- | --- |
| 0 | byte | =0x45 |
| 1 | uint16 | record count |
| 3 | minimaprec | record 1 |
| x | minimaprec | record 2 |
| ... | ... | ... |

Minimap record structure is the following:

| Offset | Size | Description |
| --- | --- | --- |
| 0 | int32 | player X coordinate |
| 4 | int32 | player Y coordinate |
| 8 | uint32 | player total mass |
| 11 | byte | player flags |
| ... | ... | ... |

player flags:

| Bit | Mask | Description |
| --- | --- | --- |
| 0 | 01 | highlight player on the minimap with red color |
| 1 | 02 |  |
| 2 | 04 |  |
| 3 | 08 |  |
| ... | ... | ... |



All player cells displayed with one minimap record, which has total mass owned by the player.
Minimap records contains top 20 players, sorted by owned mass.

This is how minimap data looks like (look for minimap debug data on the screen, because graphical minimap on the bottom right corner is not related to this feature, this is just my old minimap):

![Minimap Screenshot](https://i.imgur.com/AVTsLgK.png)

There is need more investigation about minimap record fields. If you have any info, you're welcome.

**Comment by Barbosik** [comment]
added support of server minimap to my client, works great!
Brown circles on the minimap are received from agario server :)

![Screenshot](https://i.imgur.com/NQojlnZ.png)

**Comment by SuperOP535** [comment]
ooh nice 😮

**Comment by xF4b3r** [comment]
playerX and playerY coords are int32

**Comment by ghost** [comment]
👍

**Comment by Barbosik** [comment]
> playerX and playerY coords are int32

yeah, exactly

**Comment by Barbosik** [comment]
it looks that minimap may have some small random coordinate shift and fake players. Probably in order to prevent use minimap for bots

**Comment by BorysDev** [comment]
perhaps the last byte on the minimap record is a terminator, to indicate when the record has ended

**Comment by Barbosik** [comment]
no, there is no need for terminator, because record length is known. This is some flag or player type. Needs to monitor for non zero value to understand what it means.

UPDATE: this is flags :)

00 - yellow
01 - red 
02 - yellow
03 - red

So it looks, that bit0 is used for highlight player on the minimap with red color

**Comment by Barbosik** [comment]
Here is better visual design for minimap. It shows agario server minimap with squares and local minimap with circles.

![Screenshot](https://i.imgur.com/jZbGYOH.png)

**Comment by BorysDev** [comment]
Looks good :) :+1:

**Comment by issy123** [comment]
Awesome work barbosik! Since I no longer maintain the protocol, i've invited you to be a Collaborator, this way you can organize each protocol in a seperate file and it's easier to navigate through for every agar.io clone dev

**Comment by xN3BULA** [comment]
@Barbosik Do you have a discord or something which I could contact you with other than github?

**Comment by SuperOP535** [comment]
issy I think it's better to use [Cigar2's protocol repo](https://github.com/Cigar2/Agar.io-Protocol) (it's not fully done but it has separate files and a *little* activity from the community)

**Comment by BorysDev** [comment]
But this is the original, i think we should keep it going @SuperOP535 Most of the protocol on there has been devised by Barbosik anyway. Right?

**Comment by SuperOP535** [comment]
The issues of this are original, not the repo 😂 and /shrug, we have some stuff done so like why do it again ?

**Comment by BorysDev** [comment]
@SuperOP535 If you use that laughing face unironically again i will personally hunt you down

**Comment by Mistik** [comment]
Can I use your custom client barbosik?
